File: dynamic_model.py
import numpy as np

# Default model (aerobatic plane "Zivko Edge 450")
defaultStates = [100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 900]
defaultParams = [750, 9.8056, 1.225, 9.84, 7000, 7.87, 1.25, 3531.9, 2196.4, 4887.7, 0, 0]
defaultAero = [0.05, 0.01, 0.15, -0.4, 0, 0.19, 0, 0.4, 0.1205, 5.7, -0.0002, -0.33, 0.021, -0.79, 0.075, 0, -1.23, 0, -1.1, 0, -7.34, 0.21, -0.014, -0.11, -0.024, -0.265]
defaultControls = [0, 0, 0, 0]
defaultWind = [0, 0, 0]

# Auxiliary parameters
f = 1/298.257223563  # Earth flattening
R = 6378137  # Earth equatorial radius
aux1 = 2*f-f**2
m2ft = 3.28084
rad2deg = 180/np.pi
deg2rad = np.pi/180


class Model:

	def __init__(self, states=defaultStates, params=defaultParams, aero=defaultAero, controls=defaultControls, wind=defaultWind, turbulenceIntensity=None, servosResponseTime=None, engineResponseTime=None, initVelocity=100):

		# Set initial parameters
		self.m = params[0]
		self.g = params[1]
		self.rho = params[2]
		self.S = params[3]
		self.Tmax = params[4]
		self.b = params[5]
		self.c = params[6]
		self.Ix = params[7]
		self.Iy = params[8]
		self.Iz = params[9]
		self.Ixz = params[10]
		self.incidence = params[11]*deg2rad  # Angle of incidence (degrees). Calado del ala

		# Set initial aerodynamic coefficients
		self.Cd0 = aero[0]
		self.K = aero[1]
		self.Cdb = aero[2]
		self.Cyb = aero[3]
		self.Cyda = aero[4]
		self.Cydr = aero[5]
		self.Cyp = aero[6]
		self.Cyr = aero[7]
		self.Cl0 = aero[8]
		self.Cla = aero[9]
		self.Cllb = aero[10]
		self.Cllda = aero[11]
		self.Clldr = aero[12]
		self.Cllp = aero[13]
		self.Cllr = aero[14]
		self.Cmm0 = aero[15]
		self.Cmma = aero[16]
		self.Cmmda = aero[17]
		self.Cmmde = aero[18]
		self.Cmmdr = aero[19]
		self.Cmmq = aero[20]
		self.Cnnb = aero[21]
		self.Cnnda = aero[22]
		self.Cnndr = aero[23]
		self.Cnnp = aero[24]
		self.Cnnr = aero[25]

		# Set controls
		self.da = controls[0]
		self.de = controls[1]
		self.dr = controls[2]
		self.dt = controls[3]

		# For estimation of AoA, pitch and body velocities for level flight
		self.initVelocity = initVelocity
		dynamic_pressure = 0.5 * self.rho * self.initVelocity ** 2
		term1 = np.tan((((self.m*self.g)/(dynamic_pressure*self.S))-self.Cl0)/self.Cla + self.incidence)

		# Set initial velocities
		self.vx = np.sqrt(self.initVelocity**2/(1+term1**2))
		self.vy = 0
		self.vz = self.vx * term1

		# Set internal variables
		self.V = np.sqrt(self.vx**2+self.vy**2+self.vz**2)
		self.alpha = np.arctan2(self.vz, self.vx) - self.incidence
		self.beta = np.arcsin(self.vy / self.V)
		self.Cd = self.Cy = self.Cl = 0
		self.Cll = self.Cmm = self.Cnn = 0
		self.D = self.Y = self.L = 0
		self.LL = self.MM = self.NN = 0
		self.vx_dot = self.vy_dot = self.vz_dot = 0
		self.roll_dot = self.pitch_dot = self.yaw_dot = 0
		self.p_dot = self.q_dot = self.r_dot = 0
		self.posNorth_dot = self.posEast_dot = self.alt_dot = 0
		self.lon = -3.574605617 * deg2rad  # Madrid Barajas Airport (LEMD)
		self.lat = 40.49187427 * deg2rad
		self.total_time = 0
		self.Xa = self.Ya = self.Za = 0
		self.Xt = self.Tmax*self.dt
		self.rotor_rpm = 0
		self.servosResponseTime = servosResponseTime
		self.engineResponseTime = engineResponseTime

		# Set initial states
		self.roll = 0
		self.pitch = self.alpha + self.incidence  # Same as np.arctan2(self.vz, self.vx)
		self.yaw = states[5]
		self.p = states[6]
		self.q = states[7]
		self.r = states[8]
		self.posNorth = states[9]
		self.posEast = states[10]
		self.alt = states[11]

		# Set wind
		self.windVelocity = wind[0]
		self.windHeading = wind[1] * deg2rad
		self.windElevation = wind[2] * deg2rad
		self.turbulenceIntensity = turbulenceIntensity
		self.windNorth = self.windVelocity * np.cos(self.windElevation) * np.cos(self.windHeading)
		self.windEast = self.windVelocity * np.cos(self.windElevation) * np.sin(self.windHeading)
		self.windUp = self.windVelocity * np.sin(self.windElevation)

		# Set true airspeed
		self.TAS_North = self.TAS_East = self.TAS_Up = self.TAS = self.TAS_x = self.TAS_y = self.TAS_z = 0

	def propagate(self, controls=defaultControls, dtime=1/60, mode='complete'):

		# Controls
		if self.servosResponseTime == None:
			self.da = controls[0]
			self.de = controls[1]
			self.dr = controls[2]
			self.dt = controls[3]
		else:
			self.da += (1/(60*self.servosResponseTime))*(controls[0]-self.da)
			self.de += (1/(60*self.servosResponseTime))*(controls[1]-self.de)
			self.dr += (1/(60*self.servosResponseTime))*(controls[2]-self.dr)
			self.dt += (1/(60*self.servosResponseTime))*(controls[3]-self.dt)

		# Euler trigonometry variables
		cr = np.cos(self.roll)
		sr = np.sin(self.roll)
		cp = np.cos(self.pitch)
		sp = np.sin(self.pitch)
		tp = np.tan(self.pitch)
		cy = np.cos(self.yaw)
		sy = np.sin(self.yaw)

		# Auxiliary coefficients
		coeffA = self.b/(2*self.V)
		coeffB = self.c/(2*self.V)

		# Introduce turbulence
		if self.turbulenceIntensity != None:
			turb = self.turbulenceIntensity * 10
			turbulenceCoeff = 1/np.max((np.min((-1.08*self.TAS+120, 0.1)), 100))
			self.windNorth += turbulenceCoeff*((np.random.rand()*turb-turb/2))
			self.windEast += turbulenceCoeff*((np.random.rand()*turb-turb/2))
			self.windUp += turbulenceCoeff*((np.random.rand()*turb-turb/2))

		if mode == 'complete':

			# Force coefficients
			self.Cl = self.Cl0 + self.Cla * self.alpha
			self.Cd = self.Cd0 + self.K*self.Cl**2 + self.Cdb*np.abs(self.beta)
			self.Cy = self.Cyb*self.beta + self.Cydr*self.dr + self.Cyda*self.da + coeffA*(self.Cyp*self.p+self.Cyr*self.r)

			# Moment coefficients
			self.Cll = self.Cllb*self.beta + self.Cllda*self.da + self.Clldr*self.dr + coeffA*(self.Cllp*self.p + self.Cllr*self.r)
			self.Cmm = self.Cmm0 + self.Cmma*self.alpha + self.Cmmde*self.de + self.Cmmdr*self.dr + self.Cmmda*np.abs(self.da) + coeffB*(self.Cmmq*self.q)
			self.Cnn = self.Cnnb*self.beta + self.Cnnda*self.da + self.Cnndr*self.dr + coeffA*(self.Cnnp*self.p + self.Cnnr*self.r)

		elif mode == 'longitudinal':

			# Force coefficients
			self.Cl = self.Cl0 + self.Cla * self.alpha
			self.Cd = self.Cd0 + self.K * self.Cl ** 2 + self.Cdb * np.abs(self.beta)
			self.Cy = 0

			# Moment coefficients
			self.Cll = 0
			self.Cmm = self.Cmm0 + self.Cmma * self.alpha + self.Cmmde * self.de + coeffB * (self.Cmmq * self.q)
			self.Cnn = 0

		# Auxiliary variables
		qd_times_S = 0.5*self.rho*self.TAS**2*self.S
		qd_times_S_times_b = qd_times_S*self.b
		qd_times_S_times_c = qd_times_S*self.c

		# Dynamic forces
		self.D = qd_times_S*self.Cd
		self.Y = qd_times_S*self.Cy
		self.L = qd_times_S*self.Cl

		# Dynamic moments
		self.LL = qd_times_S_times_b*self.Cll
		self.MM = qd_times_S_times_c*self.Cmm
		self.NN = qd_times_S_times_b*self.Cnn

		# Aerodynamic angles trigonometry variables
		ca = np.cos(self.alpha)
		sa = np.sin(self.alpha)
		cb = np.cos(self.beta)
		sb = np.sin(self.beta)

		# Transform forces to body-axes
		self.Xa = -ca*cb*self.D-ca*sb*self.Y+sa*self.L
		self.Ya = -sb*self.D+cb*self.Y
		self.Za = -sa*cb*self.D-sa*sb*self.Y-ca*self.L
		if self.engineResponseTime == None:
			self.Xt = self.Tmax*self.dt
		else:
			self.Xt += (1/(60*self.engineResponseTime))*(self.Tmax*self.dt-self.Xt)
		self.rotor_rpm = self.Xt*400/self.Tmax

		# Linear accelerations in body-axes
		self.vx_dot = self.r*self.vy-self.q*self.vz-self.g*sp+(self.Xa+self.Xt)/self.m
		self.vy_dot = -self.r*self.vx+self.p*self.vz+self.g*sr*cp+self.Ya/self.m
		self.vz_dot = self.q*self.vx-self.p*self.vy+self.g*cr*cp+self.Za/self.m

		# Euler rates
		self.roll_dot = self.p+tp*(self.q*sr+self.r*cr)
		self.pitch_dot = self.q*cr-self.r*sr
		self.yaw_dot = (self.q*sr+self.r*cr)/cp

		# Angular accelerations in body-axes
		aux = self.Ix*self.Iz-self.Ixz**2
		self.p_dot = (self.Ixz*(self.Ix-self.Iy+self.Iz)*self.p*self.q-(self.Iz*(self.Iz-self.Iy)+self.Ixz**2)*self.q*self.r+self.Iz*self.LL+self.Ixz*self.NN)/aux
		self.q_dot = ((self.Iz-self.Ix)*self.p*self.r-self.Ixz*(self.p**2-self.r**2)+self.MM)/self.Iy
		self.r_dot = (((self.Ix-self.Iy)*self.Ix+self.Ixz**2)*self.p*self.q-self.Ixz*(self.Ix-self.Iy+self.Iz)*self.q*self.r+self.Ixz*self.LL+self.Ix*self.NN)/aux

		# Linear velocities in NED axes
		self.posNorth_dot = self.vx*cp*cy+self.vy*(-cr*sy+sr*sp*cy)+self.vz*(sr*sy+cr*sp*cy)
		self.posEast_dot = self.vx*cp*sy+self.vy*(cr*cy+sr*sp*sy)+self.vz*(-sr*cy+cr*sp*sy)
		self.alt_dot = self.vx*sp-self.vy*sr*cp-self.vz*cr*cp

		# Propagate states
		self.vx += self.vx_dot * dtime
		self.vy += self.vy_dot * dtime
		self.vz += self.vz_dot * dtime
		self.roll += self.roll_dot * dtime
		self.pitch += self.pitch_dot * dtime
		self.yaw += self.yaw_dot * dtime
		self.p += self.p_dot * dtime
		self.q += self.q_dot * dtime
		self.r += self.r_dot * dtime
		self.posNorth += self.posNorth_dot * dtime
		self.posEast += self.posEast_dot * dtime
		self.alt += self.alt_dot * dtime

		# Transform flat earth position to LLA
		aux2 = np.sin(self.lat)**2
		RN = R/np.sqrt(1-aux1*aux2)
		RM = RN*((1-aux1)/(1-aux1*aux2))
		self.lon += np.arctan2(1, RN * np.cos(self.lat)) * self.posEast_dot * dtime
		self.lat += np.arctan2(1, RM) * self.posNorth_dot * dtime

		# Real velocity
		self.V = np.sqrt(self.vx**2+self.vy**2+self.vz**2)
		
		# True Airspeed
		self.TAS_North = self.posNorth_dot - self.windNorth
		self.TAS_East = self.posEast_dot - self.windEast
		self.TAS_Up = self.alt_dot - self.windUp
		self.TAS = np.sqrt(self.TAS_North**2 + self.TAS_East**2 + self.TAS_Up**2)
		self.TAS_x = cp*cy*self.TAS_North + cp*sy*self.TAS_East - sp*(-self.TAS_Up)  # Transform from flat-Earth axes to body-axes
		self.TAS_y = (sr*sp*cy-cr*sy)*self.TAS_North + (sr*sp*sy+cr*cy)*self.TAS_East + sr*cp*(-self.TAS_Up)
		self.TAS_z = (cr*sp*cy+sr*sy)*self.TAS_North + (cr*sp*sy-sr*cy)*self.TAS_East + cr*cp*(-self.TAS_Up)
		
		# Aerodynamic angles
		self.alpha = np.arctan2(self.TAS_z, self.TAS_x) - self.incidence
		self.beta = np.arcsin(self.TAS_y/self.TAS)

		# Accumulate time
		self.total_time += dtime

File: modules/dynamic_model.cpp
#include <dynamic_model.hpp>

//Constructor
Model::Model(float states[], float params[], float aero[],
             float controls[], float wind[], float turbulenceIntensity,
             float servosResponseTime, float engineResponseTime, float initVelocity)
{
    //Auxiliary parameters
    _f = 1.0F/298.257223563F;  //Earth flattening
    _R = 6378137;  //Earth equatorial radius
    _aux1 = 2*_f-_f*_f;
    _m2ft = 3.28084;
    _rad2deg = 180/std::_Pi;
    _deg2rad = std::_Pi/180;

    //Set initial parameters
    _m = params[0];
    _g = params[1];
    _rho = params[2];
    _S = params[3];
    _Tmax = params[4];
    _b = params[5];
    _c = params[6];
    _Ix = params[7];
    _Iy = params[8];
    _Iz = params[9];
    _Ixz = params[10];
    _incidence = params[11] * _deg2rad; //Angle of incidence (degrees). Calado del ala

    //Set initial aerodynamic coefficients
    _Cd0 = aero[0];
    _K = aero[1];
    _Cdb = aero[2];
    _Cyb = aero[3];
    _Cyda = aero[4];
    _Cydr = aero[5];
    _Cyp = aero[6];
    _Cyr = aero[7];
    _Cl0 = aero[8];
    _Cla = aero[9];
    _Cllb = aero[10];
    _Cllda = aero[11];
    _Clldr = aero[12];
    _Cllp = aero[13];
    _Cllr = aero[14];
    _Cmm0 = aero[15];
    _Cmma = aero[16];
    _Cmmda = aero[17];
    _Cmmde = aero[18];
    _Cmmdr = aero[19];
    _Cmmq = aero[20];
    _Cnnb = aero[21];
    _Cnnda = aero[22];
    _Cnndr = aero[23];
    _Cnnp = aero[24];
    _Cnnr = aero[25];

    //Set controls
    _da = controls[0];
    _de = controls[1];
    _dr = controls[2];
    _dt = controls[3];

    //For estimation of AoA, pitch and body velocities for level flight
    _initVelocity = initVelocity;
    float dynamic_pressure = 0.5 * _rho * _initVelocity * _initVelocity;
    float term1 = tanf((((_m*_g)/(dynamic_pressure*_S))-_Cl0)/_Cla + _incidence);

    //Set initial velocities
    _vx = sqrtf(_initVelocity*_initVelocity/(1+term1*term1));
    _vy = 0.0F;
    _vz = _vx * term1;

    //Set internal variables
    _V = sqrtf(_vx*_vx+_vy*_vy+_vz*_vz);
    _alpha = atan2f(_vz, _vx) - _incidence;
    _beta = asin(_vy / _V);
    _Cd = 0.0F;
    _Cy = 0.0F;
    _Cl = 0.0F;
    _Cll = 0.0F;
    _Cmm = 0.0F;
    _Cnn = 0.0F;
    _D = 0.0F;
    _Y = 0.0F;
    _L = 0.0F;
    _LL = 0.0F;
    _MM = 0.0F;
    _NN = 0.0F;
    _vx_dot = 0.0F;
    _vy_dot = 0.0F;
    _vz_dot = 0.0F;
    _roll_dot = 0.0F;
    _pitch_dot = 0.0F;
    _yaw_dot = 0.0F;
    _p_dot = 0.0F;
    _q_dot = 0.0F;
    _r_dot = 0.0F;
    _posNorth_dot = 0.0F;
    _posEast_dot = 0.0F;
    _alt_dot = 0.0F;
    _lon = -3.574605617 * _deg2rad;  //Madrid Barajas Airport (LEMD)
    _lat = 40.49187427 * _deg2rad;
    _total_time = 0.0F;
    _Xa = 0.0F;
    _Ya = 0.0F;
    _Za = 0.0F;
    _Xt = _Tmax*_dt;
    _rotor_rpm = 0.0F;
    _servosResponseTime = servosResponseTime;
    _engineResponseTime = engineResponseTime;

    //Set initial states
    _roll = 0.0F;
    _pitch = _alpha + _incidence;  //Same as np.arctan2(_vz, _vx)
    _yaw = states[5];
    _p = states[6];
    _q = states[7];
    _r = states[8];
    _posNorth = states[9];
    _posEast = states[10];
    _alt = states[11];

    //Set wind
    _windVelocity = wind[0];
    _windHeading = wind[1] * _deg2rad;
    _windElevation = wind[2] * _deg2rad;
    _turbulenceIntensity = turbulenceIntensity;
    _windNorth = _windVelocity * cosf(_windElevation) * cosf(_windHeading);
    _windEast = _windVelocity * cosf(_windElevation) * sinf(_windHeading);
    _windUp = _windVelocity * sinf(_windElevation);

    //Set true airspeed
    _TAS_North = 0.0F;
    _TAS_East = 0.0F;
    _TAS_Up = 0.0F;
    _TAS = 0.0F;
    _TAS_x = 0.0F;
    _TAS_y = 0.0F;
    _TAS_z = 0.0F;
}

void Model::propagate(float controls[], float dtime)
{
    //Controls
    if (_servosResponseTime == 0.0F)
    {
        _da = controls[0];
        _de = controls[1];
        _dr = controls[2];
        _dt = controls[3];
    }
    else
    {
        _da += (1.0F/(60.0F*_servosResponseTime))*(controls[0]-_da);
        _de += (1.0F/(60.0F*_servosResponseTime))*(controls[1]-_de);
        _dr += (1.0F/(60.0F*_servosResponseTime))*(controls[2]-_dr);
        _dt += (1.0F/(60.0F*_servosResponseTime))*(controls[3]-_dt);
    }

    //Euler trigonometry variables
    float cr = cosf(_roll);
    float sr = sinf(_roll);
    float cp = cosf(_pitch);
    float sp = sinf(_pitch);
    float tp = tan(_pitch);
    float cy = cosf(_yaw);
    float sy = sinf(_yaw);

    //Auxiliary coefficients
    float coeffA = _b/(2.0F*_V);
    float coeffB = _c/(2.0F*_V);

    //Introduce turbulence
    if (_turbulenceIntensity != 0.0F)
    {
        std::mt19937 _gen;
        std::uniform_real_distribution<float> _dist(0, 1);
        float turb = _turbulenceIntensity * 10.0F;
        float turbulenceCoeff = 1.0F/std::max(std::min(-1.08F*_TAS+120.0F, 0.1F), 100.0F);
        _windNorth += turbulenceCoeff * (_dist(_gen)*turb - turb/2.0F);
        _windEast += turbulenceCoeff * (_dist(_gen)*turb - turb/2.0F);
        _windUp += turbulenceCoeff * (_dist(_gen)*turb - turb/2.0F);
    }

    //Force coefficients
    _Cl = _Cl0 + _Cla * _alpha;
    _Cd = _Cd0 + _K * _Cl * _Cl + _Cdb * std::abs(_beta);
    _Cy = _Cyb*_beta + _Cydr*_dr + _Cyda*_da + coeffA*(_Cyp*_p+_Cyr*_r);

    //Moment coefficients
    _Cll = _Cllb*_beta + _Cllda*_da + _Clldr*_dr + coeffA*(_Cllp*_p + _Cllr*_r);
    _Cmm = _Cmm0 + _Cmma*_alpha + _Cmmde*_de + _Cmmdr*_dr + _Cmmda*std::abs(_da) + coeffB*(_Cmmq*_q);
    _Cnn = _Cnnb*_beta + _Cnnda*_da + _Cnndr*_dr + coeffA*(_Cnnp*_p + _Cnnr*_r);

    //Auxiliary variables
    float qd_times_S = 0.5*_rho*_TAS*_TAS*_S;
    float qd_times_S_times_b = qd_times_S*_b;
    float qd_times_S_times_c = qd_times_S*_c;

    //Dynamic forces
    _LL = qd_times_S_times_b*_Cll;
    _MM = qd_times_S_times_c*_Cmm;
	_NN = qd_times_S_times_b*_Cnn;

    //Dynamic moments
    _LL = qd_times_S_times_b*_Cll;
    _MM = qd_times_S_times_c*_Cmm;
    _NN = qd_times_S_times_b*_Cnn;

    //Aerodynamic angles trigonometry variables
    float ca = cosf(_alpha);
    float sa = sinf(_alpha);
    float cb = cosf(_beta);
    float sb = sinf(_beta);

    //Transform forces to body-axes
    _Xa = -ca*cb*_D-ca*sb*_Y+sa*_L;
    _Ya = -sb*_D+cb*_Y;
    _Za = -sa*cb*_D-sa*sb*_Y-ca*_L;
    if (_engineResponseTime == 0.0F)
    {
        _Xt = _Tmax*_dt;
    }
    else
    {
        _Xt += (1/(60*_engineResponseTime))*(_Tmax*_dt-_Xt);
    }
    _rotor_rpm = _Xt*400/_Tmax;

    //Linear accelerations in body-axes
    _vx_dot = _r*_vy-_q*_vz-_g*sp+(_Xa+_Xt)/_m;
    _vy_dot = -_r*_vx+_p*_vz+_g*sr*cp+_Ya/_m;
    _vz_dot = _q*_vx-_p*_vy+_g*cr*cp+_Za/_m;

    //Euler rates
    _roll_dot = _p+tp*(_q*sr+_r*cr);
    _pitch_dot = _q*cr-_r*sr;
    _yaw_dot = (_q*sr+_r*cr)/cp;

    //Angular accelerations in body-axes
    float aux = _Ix*_Iz-_Ixz*_Ixz;
    _p_dot = (_Ixz*(_Ix-_Iy+_Iz)*_p*_q-(_Iz*(_Iz-_Iy)+_Ixz*_Ixz)*_q*_r+_Iz*_LL+_Ixz*_NN)/aux;
    _q_dot = ((_Iz-_Ix)*_p*_r-_Ixz*(_p*_p-_r*_r)+_MM)/_Iy;
    _r_dot = (((_Ix-_Iy)*_Ix+_Ixz*_Ixz)*_p*_q-_Ixz*(_Ix-_Iy+_Iz)*_q*_r+_Ixz*_LL+_Ix*_NN)/aux;

    //Linear velocities in NED axes
    _posNorth_dot = _vx*cp*cy+_vy*(-cr*sy+sr*sp*cy)+_vz*(sr*sy+cr*sp*cy);
    _posEast_dot = _vx*cp*sy+_vy*(cr*cy+sr*sp*sy)+_vz*(-sr*cy+cr*sp*sy);
    _alt_dot = _vx*sp-_vy*sr*cp-_vz*cr*cp;

    //Propagate states
    _vx += _vx_dot * dtime;
    _vy += _vy_dot * dtime;
    _vz += _vz_dot * dtime;
    _roll += _roll_dot * dtime;
    _pitch += _pitch_dot * dtime;
    _yaw += _yaw_dot * dtime;
    _p += _p_dot * dtime;
    _q += _q_dot * dtime;
    _r += _r_dot * dtime;
    _posNorth += _posNorth_dot * dtime;
    _posEast += _posEast_dot * dtime;
    _alt += _alt_dot * dtime;

    //Transform flat earth position to LLA
    float aux2 = sinf(_lat);
    aux2 *= aux2;
    float RN = _R/sqrtf(1-_aux1*aux2);
    float RM = RN*((1-_aux1)/(1-_aux1*aux2));
    _lon += atan2f(1, RN * cosf(_lat)) * _posEast_dot * dtime;
    _lat += atan2f(1, RM) * _posNorth_dot * dtime;

    //Real velocity
    _V = sqrtf(_vx*_vx+_vy*_vy+_vz*_vz);
    
    //True Airspeed
    _TAS_North = _posNorth_dot - _windNorth;
    _TAS_East = _posEast_dot - _windEast;
    _TAS_Up = _alt_dot - _windUp;
    _TAS = sqrtf(_TAS_North*_TAS_North + _TAS_East*_TAS_East + _TAS_Up*_TAS_Up);
    _TAS_x = cp*cy*_TAS_North + cp*sy*_TAS_East - sp*(-_TAS_Up);  //Transform from flat-Earth axes to body-axes
    _TAS_y = (sr*sp*cy-cr*sy)*_TAS_North + (sr*sp*sy+cr*cy)*_TAS_East + sr*cp*(-_TAS_Up);
    _TAS_z = (cr*sp*cy+sr*sy)*_TAS_North + (cr*sp*sy-sr*cy)*_TAS_East + cr*cp*(-_TAS_Up);
    
    //Aerodynamic angles
    _alpha = atan2f(_TAS_z, _TAS_x) - _incidence;
    _beta = asinf(_TAS_y/_TAS);

    //Accumulate time
    _total_time += dtime;
}